crates/fortitude_preprocessor/src/cpp.rs

```diff
@@ -0,0 +1,852 @@
+use crate::cpp_tokens::{CppDirectiveKind, CppTokenIterator, CppTokenKind};
+use anyhow::{Context, anyhow};
+use chrono::prelude::*;
+use lazy_regex::regex_captures;
+use ruff_source_file::{OneIndexed, SourceCode};
+use ruff_text_size::TextSize;
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+
+/// Enum describing where a snippet came from.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Provenance {
+    /// A system-defined macro, e.g. __DATE__ or __TIME__.
+    SystemDefined,
+    /// A user-defined macro, e.g. passed in via command line.
+    UserDefined,
+    /// A macro defined in the source file.
+    FileDefined {
+        start: usize,
+        end: usize,
```

Comment:
Is the plan to switch to `ruff_text_size::TextRange`?

Reply:
Yeah, they're definitely a bit more annoying, but `TextRange` is 1/4 the size of `(usize, usize)`, and if we're going to make a whole bunch of these, it's probably worth it.

But definitely do the simple thing and get it working first!